scripts/storybook/utils.js

```diff
@@ -0,0 +1,86 @@
+// @ts-check
+
+const fs = require('fs');
+const path = require('path');
+const { stripIndents } = require('@nrwl/devkit');
+const { workspaceRoot } = require('nx/src/utils/app-root');
+
+const defaults = { workspaceRoot };
+/**
+ * Registers workspace custom storybook addon to storybook with build-less setup during development.
+ *
+ * For production builds this behaves as identity function (returns the passed addonName)
+ *
+ * @example
+ * ```js
+ *  module.exports = { addons: ['@storybook/addon-essentials', loadWorkspacePlugin('@fluentui/custom-storybook-addon')] }
+ * ```
+ *
+ * @param {string} addonName - package name of custom workspace addon
+ * @param {{workspaceRoot?:string}=} options
+ */
+function loadWorkspacePlugin(addonName, options = {}) {
+  const { workspaceRoot } = { ...defaults, ...options };
+
+  if (process.env.NODE_ENV === 'production') {
+    return addonName;
+  }
+
+  function getPaths() {
+    const workspaceJson = JSON.parse(fs.readFileSync(path.join(workspaceRoot, 'workspace.json'), 'utf-8'));
+    const addonMetadata = workspaceJson.projects[addonName];
+    const rootPath = path.join(workspaceRoot, addonMetadata.root);
+    const sourceRootPath = path.join(workspaceRoot, addonMetadata.sourceRoot);
+    const tsConfigPath = path.join(rootPath, 'tsconfig.lib.json');
+    const packageJsonPath = path.join(rootPath, 'package.json');
+    /**
+     * @type {Record<string,any>}
+     */
+    const tsconfigLibJson = JSON.parse(fs.readFileSync(tsConfigPath, 'utf-8'));
+    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
+    const packageDistPath = path.dirname(packageJson.module);
+    const distPath = path.join(rootPath, tsconfigLibJson.compilerOptions.outDir);
+    const relativePathToSource = path.relative(distPath, sourceRootPath);
+    const presetSourcePath = path.join(rootPath, 'preset.js');
+    const presetMemorySourcePath = path.join(distPath, 'preset.js');
+
+    return {
+      distPath,
+      packageDistPath,
+      presetSourcePath,
+      presetMemorySourcePath,
+      relativePathToSource,
+    };
+  }
+
+  const { relativePathToSource, distPath, packageDistPath, presetSourcePath, presetMemorySourcePath } = getPaths();
+
+  if (!fs.existsSync(presetSourcePath)) {
+    throw new Error(
+      'Looks like your storybook addon is missing "preset.js" module. This file needs to be placed on root of your package',
+    );
+  }
+
+  const presetContent = fs.readFileSync(presetSourcePath, 'utf-8');
+
+  const regex = new RegExp(`\.\\/${path.normalize(packageDistPath)}`, 'g');
+  let modifiedPresetContent = presetContent.replace(regex, relativePathToSource);
+  modifiedPresetContent = stripIndents`
+    const { workspaceRoot } = require('nx/src/utils/app-root');
+    const { registerTsProject } = require('nx/src/utils/register');
```

Comment:
what does `registerTsProject` do exactly ? I can't find docs for it. I'm not sure why `dist` path needs to be used if it builds in memory.

Reply:
good questions! I was hoping the e2e test will provide the answers :D lemme spin up a quick README.md